drake/solvers/Optimization.h

```diff
@@ -0,0 +1,323 @@
+
+      virtual bool solve(OptimizationProblem& prog) { throw std::runtime_error("not implemented yet"); }
+    };
+
+/*  // Prototype of the more complete optimization problem class hiearchy (to be implemented only as needed)
+    struct MixedIntegerNonlinearProgram : public MathematicalProgram {};
+    struct MixedIntegerSemidefiniteProgram : public MixedIntegerNonlinearProgram {};
+    struct MixedIntegerSecondOrderConeProgram : public MixedIntegerSemidefiniteProgram {};
+    struct MixedIntegerQuadraticProgram : public MixedIntegerSecondOrderConeProgram {};
+    struct MixedIntegerLinearProgram : public MixedIntegerQuadraticProgram {};
+
+    struct NonlinearProgram : public MixedIntegerNonlinearProgram {};
+    struct SemidefiniteProgram : public NonlinearProgram, public MixedIntegerSemidefiniteProgram {};
+    struct SecondOrderConeProgram : public SemidefiniteProgram, public MixedIntegerSecondOrderConeProgram {};
+    struct QuadraticProgram : public SecondOrderConeProgram, public MixedIntegerQuadraticProgram {};
+    struct LinearProgram : public QuadraticProgram, public MixedIntegerLinearProgram {
+      virtual MathematicalProgram* addLinearEqualityConstraint() { return new LinearProgram; };
+      virtual MathematicalProgram* addLinearInequalityConstraint() { return new LinearProgram; };
+    };
+
+    struct NonlinearComplementarityProblem : public NonlinearProgram {};
+    struct LinearComplementarityProblem : public NonlinearComplementarityProblem {};
+*/
+    struct LeastSquares : public MathematicalProgram { //public LinearProgram, public LinearComplementarityProblem {
+      virtual MathematicalProgram* addLinearEqualityConstraint() { return new LeastSquares; };
+      virtual bool solve(OptimizationProblem& prog) {
+        size_t num_constraints = 0;
+        for (auto const& c : prog.linear_equality_constraints) {
+          num_constraints += c->getMatrix().rows();
+        }
+
+        Eigen::MatrixXd Aeq = Eigen::MatrixXd::Zero(num_constraints,prog.num_vars);  // todo: use a sparse matrix here?
+        Eigen::VectorXd beq(num_constraints);
+
+        size_t constraint_index = 0;
+        for (auto const& c : prog.linear_equality_constraints) {
+          size_t n = c->getMatrix().rows();
+          size_t var_index = 0;
+          for (const DecisionVariableView& v : c->getVariableList() ) {
+            Aeq.block(constraint_index, v.index(), n, v.size()) = c->getMatrix().middleCols(var_index, v.size());
+            var_index += v.size();
+          }
+          beq.segment(constraint_index, n) = c->getVector();  // = c->getUpperBound() since it's an equality constraint
+          constraint_index += n;
+        }
+
+        // least-squares solution
+        prog.setSolution(Aeq.jacobiSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(beq));
+        return true;
+      }
+    };
+    std::shared_ptr<MathematicalProgram> problem_type;
```

Comment:
If we only support moving up, then the class hierarchy requires every parent class can have only one child class, but we can support multiple inheritance, is this right? So MixedIntegerNonlinearProgram cannot have two child classes, as NonlinearProgram and IntegerNonlinearProgram. 

Why `LeastSquares` should be the most derived class? I thought it should be a problem with integer objective and equality integer constraint. Or do you plan to change the most derived class later when we encounter such problems?